### ISSUE TITLE
ECJ omits synthetic parameters from RuntimeIn/VisibleAnnotations array

### ISSUE THREAD

**Issue by smowton** [eclipse-jdt/eclipse.jdt.core#2787]
Consider input program:

Test.java:
```java
public enum Test {

  A("A"), B("B"), C("C");

  private Test(@Ann String x) { }

}

@interface Ann {}
```

Test2.java:
```java
public class Test2 { Test t; }
```

If the former is built with ECJ and the latter with `javac`, like:

```
ecj Test.java -source 8 -d out
javac Test2.java -cp out
```

Then `ecj` will emit a constructor for `Test` with method signature `(int, String, String)` (adding two synthetic parameters for the enum ordinal and description), but a `RuntimeInvisibleAnnotations` array of length 1 (omitting the synthetic parameters).

`javac` will hit an array OOB due to this and thus throw a `BadClassFile` error.

The JVMS as of Java 9 says at https://docs.oracle.com/javase/specs/jvms/se21/html/jvms-4.html#jvms-4.7.17 that the annotations array does not have to declare exactly one entry per parameter, but `javac` has never actually respected this (https://bugs.openjdk.org/browse/JDK-8334871 intends to remedy this in Java 24, but older Javas will continue to have problems with `ecj`-generated class files).

By ensuring the generated array always matches the actual method parameter list length (including synthetics), you would improve compatibility with `javac` and `javac`-derived tooling.

**Comment by stephan-herrmann** [comment]
Thanks for researching the background before reporting here.

> By ensuring the generated array always matches the actual method parameter list length (including synthetics), you would improve compatibility with `javac` and `javac`-derived tooling.

While we might improve compatibility at one end, the same change could also break existing clients. Annotation processors or reflection based tools may not expect to see annotations on parameters that don't exist in source code.

I don't think such a behavioral change would be wise at this point.

Incompatibility is a natural consequence of a loose specification. And incompatibility between different versions of ecj is problematic on its own right.

**Comment by smowton** [comment]
Given `javac` itself generates annotation arrays of the "right" size (according to the JVMS as at JDK 8), any tooling out there that was *more* surprised by `javac`'s behaviour than ECJ's would be exclusively operating in an Eclipse ecosystem.

Re: ECJ vs. ECJ compatibility, I note that compiling `Test.java` using `javac` and then the user `Test2.java` using ECJ works at present, which suggests that as expected ECJ is at least willing to tolerate an annotation array including every parameter in the method signature, synthetic or otherwise.

**Comment by stephan-herrmann** [comment]
> Given `javac` itself generates annotation arrays of the "right" size (according to the JVMS as at JDK 8),

That's not how I read [JDK-8334871](https://bugs.openjdk.org/browse/JDK-8334871), which says:
> "javac was not implementing the specification correctly in two aspects:
> *  when writing a classfile, when the Signature attribute was used for a method, it was only producing as many entries of these attributes as there were parameters in the Signature attribute [...]"

javac had a bug at 1.8, which was legalized only later.

> any tooling out there that was _more_ surprised by `javac`'s behaviour than ECJ's would be exclusively operating in an Eclipse ecosystem.

I think we should first pay attention not to break clients in our own ecosystem, before thinking of other clients, shouldn't we?
 
> Re: ECJ vs. ECJ compatibility, I note that compiling `Test.java` using `javac` and then the user `Test2.java` using ECJ works at present, which suggests that as expected ECJ is at least willing to tolerate an annotation array including every parameter in the method signature, synthetic or otherwise.

That's not the use case I had in mind. I meant to speak (again) of the situation of clients observing different behavior from different versions of ecj. This shouldn't happen other than as a bug fix nor in reaction to a spec change. 

Not throwing exceptions on ecj's legally produced class files is something that javac can fix and backport as many levels back as they wish, **without causing any incompatibility**. ecj has no such option in this regard.

**Comment by stephan-herrmann** [comment]
Forgot to mention: at the bottom of the problem is a disagreement regarding Signature attributes: javac generates and expects those to be present in constructors of enums, while JVMS has no indication that those should be generated for enums. See  https://github.com/eclipse-jdt/eclipse.jdt.core/issues/205 and its outgoing links.

**Comment by smowton** [comment]
> That's not how I read [JDK-8334871](https://bugs.openjdk.org/browse/JDK-8334871), which says...

I think we're at cross purposes there -- I believe we agree that old `javac` produced a "short" annotation array, ECJ does the same, current `javac` produces (and demands) a "long" array, and in particular it cannot tolerate ECJ's output. It sounds like it might tolerate the old `javac` output due to its use of the `Signature` attribute.

Agreed that there's a possibility to surprise any consumers that have already worked around the current javac/ecj disagreement, or that expect exclusively ecj class files. Nonetheless it would seem useful to be able to produce a class file that doesn't break current `javac`, considering their fix isn't scheduled until two major versions in the future, and doesn't read like a high-priority bug so may well slip further. Do you ever allow command-line flags for situations like this, so one could explicitly ask ECJ to produce `javac`-compatible otuput?

**Comment by stephan-herrmann** [comment]
> > That's not how I read [JDK-8334871](https://bugs.openjdk.org/browse/JDK-8334871), which says...
> 
> I think we're at cross purposes there -- I believe we agree that old `javac` produced a "short" annotation array, ECJ does the same, current `javac` produces (and demands) a "long" array, and in particular it cannot tolerate ECJ's output. It sounds like it might tolerate the old `javac` output due to its use of the `Signature` attribute.

Yes, the `Signature` attribute might well be the reason. Summary: javac has two bugs (not accepting spec-conforming byte code, plus generating an unspecified Signature attribute), which seem to partially compensate each other.

> Agreed that there's a possibility to surprise any consumers that have already worked around the current javac/ecj disagreement, or that expect exclusively ecj class files. Nonetheless it would seem useful to be able to produce a class file that doesn't break current `javac`, considering their fix isn't scheduled until two major versions in the future, and doesn't read like a high-priority bug so may well slip further. Do you ever allow command-line flags for situations like this, so one could explicitly ask ECJ to produce `javac`-compatible output?

We don't have the resources to add complexity to ecj (which must be maintained for years to come) just to please buggy javac. Two things you could do: 
* instead of arguing with us put pressure on Oracle to fix the javac bugs sooner rather than later, or
* contribute a high quality patch for ecj that introduces the proposed option switch.
